[SCRUM-110] ✨ BE 테스트 데이터베이스 연동

GideokKim opened the pull request:
## Explain this Pull Request 🙏

- Jira URL: https://9git-9git.atlassian.net/browse/SCRUM-110

## What has fixed? 🛠

변경 전 내용

변경 후 내용

- [Azure DB 리소스](https://portal.azure.com/#@msaischool.onmicrosoft.com/resource/subscriptions/b850d62a-25fe-4d3a-9697-ea40449528a9/resourceGroups/gugit-rsg/providers/Microsoft.DBforPostgreSQL/flexibleServers/gugit-test-db/overview)와 연동 완료

## Whether needed review 🖍

<!-- 마크다운의 빈 체크박스 문법은 [ ] / 채워진 체크박스 문법은 [x] 입니다. -->

- [x] 🙋🏻 리뷰가 필요합니다!
- [ ] 🙅🏻 리뷰가 필요하지 않습니다!

(이유 : )

## Document🗒️ (Optional)

- 로컬에서 테스트하려면 `.env` 파일이 필요합니다. 필요할 경우 @GideokKim 에게 요청하세요.

## Screenshot 📸 (Optional)

## Test Checklist ✅ (Optional)

- 아래 명령어 순차적으로 실행

```bash
poetry install --no-root
```

```bash
poetry run uvicorn app.main:app --reload
```

<!-- 마크다운의 빈 체크박스 문법은 [ ] / 채워진 체크박스 문법은 [x] 입니다. -->

GideokKim committed:
📝 프로젝트 세팅 단계 명령어 수정

GideokKim committed:
✨ 데이터베이스 세션 관리 구현

GideokKim committed:
📝 `poetry` 가이드라인 작성

GideokKim committed:
📝 `uvicorn` 가이드라인 작성

GideokKim committed:
📝 폴더 구조 및 명령어 수정

GideokKim committed:
🔧 Authors 추가

GideokKim committed:
📝 `FastAPI` 가이드라인 작성

GideokKim committed:
📝 기술스택 관련 내용 추가

GideokKim committed:
✨ DB 초기화 추가

GideokKim committed:
➕ `pytz` 종속성 추가

GideokKim committed:
🐛 `SQLAlchemy`의 `UUID` 타입을 사용하도록 수정

Ryan-OH reviewed (approved):
LGTM

trivialcoding6 commented:
@GideokKim 
저는 아래 모듈이 없어서 한번 에러 났었는데 다른 분들은 잘 실행되셨나용!?

@codingorijazz @Ryan-OH @chaeengg 
<img width="813" alt="스크린샷 2025-04-08 오전 9 41 47" src="https://github.com/user-attachments/assets/ed5a5134-be56-4b47-b146-ab649b0578be" />

trivialcoding6 commented on app/db/base.py:
```diff
@@ -0,0 +1,34 @@
+
+    try:
+        async with engine.begin() as conn:
+            await conn.run_sync(Base.metadata.create_all)
```
@GideokKim 
여기에서 Base.metadata.create_all 해주는데 저희가 Base에 메타데이터를 깜빡한거 같아서 아래대로 models/base.py 에 추가하면 될 거 같습니다!!

```
from sqlalchemy import DateTime, func, MetaData

class Base(DeclarativeBase):
    """Base class which provides automated table name
    and surrogate primary key column.
    """

    metadata = MetaData()
```

GideokKim replied:
수정 이후 다른 곳에서도 에러가 발생하면서 테이블 생성이 안되었습니다 ㅠㅠ 관련해서 수정 후 push했습니다!

GideokKim committed:
📝 Azure data studio 설치 및 데이터베이스 연동 가이드라인 작성

GideokKim committed:
🔊 데이터베이스 연동 관련 로그 추가

GideokKim committed:
🐛 누락된 `metadata` 코드 추가

`metadata`는 DB 스키마 정보를 담는 컨테이너이므로 없으면 안됨.

GideokKim committed:
💥 모든 model 파일에 대한 임포트 추가

모든 모델 클래스가 메모리에 로드되지 않으면 생성되지 않으므로 임포트가 필요함.

GideokKim committed:
🔥 `UUID(as_uuid=True)` 제거

이미 `Mapped[UUID]`에서 UUID가 들어가기 때문에 불필요한 코드였고, 중복 지정하면 테이블 만들때 에러가 발생함.

GideokKim committed:
🚑️ 모델 간 순환 참조 문제 해결

`user.py`가 다른 모델 코드를 참조하고 다른 모델 코드는 `user.py`를 참조하면서 어떤 테이블을 먼저 만들지 머신이 판단 불가능한 상태였음.

GideokKim committed:
🚑️ 모델 간 순환 참조 문제 해결

`category.py`와 `evaluation.py`가 서로 참조하면서 어떤 테이블을 먼저 만들지 머신이 판단 불가능한 상태였음.

GideokKim committed:
🚑️ 잘못된 `__table_args__` 사용 수정

GideokKim committed:
✏️ 단수 및 복수 철자 오타 수정

GideokKim committed:
🚑️ 잘못된 관계 설정 수정

trivialcoding6 commented on app/models/user.py:
```diff
@@ -33,26 +30,38 @@ class User(Base):
     )
-    goals: Mapped[List["Goal"]] = relationship(
-        "Goal", back_populates="user", cascade="all, delete"
+    goals: Mapped[List["app.models.category.Goal"]] = relationship(
```
@GideokKim 
혹시 요거 수정하실 때 참고하신 문서가 있을까요!? 다른건 아니고 어떻게 해결할지 검색해보다가 아래 링크를 보았는데 링크랑은 다르게 아예 import 경로를 넣으신 방법으로 해결하신거 같아서 
아래는 그냥 블로그라 공신력은 없는거 같고 혹시 좀 더 공식적으로 확인할 수 있는 곳이 있나 궁금했습니닷

https://velog.io/@silica_o3o/FastAPI-%EC%88%9C%ED%99%98-%EC%B0%B8%EC%A1%B0-%ED%95%B4%EA%B2%B0%ED%95%98%EA%B8%B0

GideokKim replied:
@trivialcoding6 순환 참조 문제는 FastAPI가 아니라 파이썬 자체의 모듈 시스템 특성으로 인한 문제라고 생각했습니다!

SQLAlchemy 공식 문서 중 [Late-Evaluation of Relationship Arguments](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#late-evaluation-of-relationship-arguments)에 따르면 SQLAlchemy에는 로딩 지연 메커니즘이 있어서 파이썬 자체의 모듈 시스템 특성을 피하면서 순환 참조 문제를 해결할 수 있을 것 같아서 이렇게 작성했습니다!

링크주신 블로그 글 보니 여러 해결 방법이 있는데, 저 해결 방법을 저는 모르고 있었어서 사용하지 못했네요 ㅠㅠ 그래도 블로그 글에서도 마지막 부분에 제가 선택한 방법에 대한 언급도 있는듯합니당

trivialcoding6 commented:
LGTM
위 코멘트는 단순 질문이었어서 기덕님이 확인하시고 resolve 처리해주시면 될 거 같습니다!!
(테이블 생성되는것은 확인했습니다!)
👍👍👍👍
